File: lib/domain/entities/mosque.dart
import 'package:equatable/equatable.dart';
import 'prayer_times.dart';

class Mosque extends Equatable {
  final String id;
  final String name;
  final String? address;
  final String? city;
  final String? country;
  final double? latitude;
  final double? longitude;
  final bool isFavorite;
  final DateTime? lastAccessed;

  const Mosque({
    required this.id,
    required this.name,
    this.address,
    this.city,
    this.country,
    this.latitude,
    this.longitude,
    this.isFavorite = false,
    this.lastAccessed,
  });

  Mosque copyWith({
    String? id,
    String? name,
    String? address,
    String? city,
    String? country,
    double? latitude,
    double? longitude,
    bool? isFavorite,
    DateTime? lastAccessed,
  }) {
    return Mosque(
      id: id ?? this.id,
      name: name ?? this.name,
      address: address ?? this.address,
      city: city ?? this.city,
      country: country ?? this.country,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      isFavorite: isFavorite ?? this.isFavorite,
      lastAccessed: lastAccessed ?? this.lastAccessed,
    );
  }

  @override
  List<Object?> get props => [
    id, name, address, city, country,
    latitude, longitude, isFavorite, lastAccessed,
  ];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'address': address,
      'city': city,
      'country': country,
      'latitude': latitude,
      'longitude': longitude,
      'isFavorite': isFavorite,
      'lastAccessed': lastAccessed?.toIso8601String(),
    };
  }

  factory Mosque.fromJson(Map<String, dynamic> json) {
    return Mosque(
      id: json['id'] as String,
      name: json['name'] as String,
      address: json['address'] as String?,
      city: json['city'] as String?,
      country: json['country'] as String?,
      latitude: json['latitude'] != null ? (json['latitude'] as num).toDouble() : null,
      longitude: json['longitude'] != null ? (json['longitude'] as num).toDouble() : null,
      isFavorite: json['isFavorite'] as bool? ?? false,
      lastAccessed: json['lastAccessed'] != null 
        ? DateTime.parse(json['lastAccessed'] as String) 
        : null,
    );
  }
}

File: src-tauri/src/models/mosque.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::PrayerTimes;

/// Mosque entity
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Mosque {
    pub id: String,
    pub name: String,
    pub address: Option<String>,
    pub city: Option<String>,
    pub country: Option<String>,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    pub is_favorite: bool,
    pub last_accessed: Option<DateTime<Utc>>,
}

impl Mosque {
    pub fn new(id: String, name: String) -> Self {
        Self {
            id,
            name,
            address: None,
            city: None,
            country: None,
            latitude: None,
            longitude: None,
            is_favorite: false,
            last_accessed: None,
        }
    }
}

/// Mosque with prayer times
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MosqueWithPrayerTimes {
    #[serde(flatten)]
    pub mosque: Mosque,
    pub prayer_times: Option<PrayerTimes>,
}

/// Search result for mosques
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MosqueSearchResult {
    pub mosques: Vec<Mosque>,
    pub total: usize,
}

/// Favorite mosque entry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FavoriteMosque {
    pub mosque: Mosque,
    pub added_at: DateTime<Utc>,
    pub is_active: bool,
}
